frontend/leaderboard/src/Challenges.re

```diff
@@ -10,55 +10,128 @@ let printMap = map => {
+let getBlocksCreatedByUser = blocks => {
+  blocks
+  |> Array.fold_left(
+       (map, block: Types.NewBlock.t) => {
+         updateMapValue(block.data.newBlock.creatorAccount.publicKey, map)
+       },
+       StringMap.empty,
+     );
 };
 
-// Expected Output
-// {
-//       "pk1": {"block_count": 1, "transactions_sent": 134, "snark_jobs": 11}
-//       "pk2": {"block_count": 4, "transactions_sent": 55, "snark_jobs": 3}
-//       "pk3": {"block_count": 0, "transactions_sent": 3, "snark_jobs": 8}
-//}
-let handleMetrics = (metrics, blocks) => {
+let calculateTransactionSent = (map, block: Types.NewBlock.t) => {
+  block.data.newBlock.transactions.userCommands
+  |> Array.fold_left(
+       (transactionMap, userCommand: Types.NewBlock.userCommands) => {
+         updateMapValue(userCommand.fromAccount.publicKey, transactionMap)
+       },
+       map,
+     );
+};
+
+let getTransactionSentByUser = blocks => {
+  blocks |> calculateProperty(calculateTransactionSent);
+};
+
+let calculateSnarkWorkCount = (map, block: Types.NewBlock.t) => {
+  block.data.newBlock.snarkJobs
+  |> Array.fold_left(
+       (snarkMap, snarkJob: Types.NewBlock.snarkJobs) => {
+         updateMapValue(snarkJob.prover, snarkMap)
+       },
+       map,
+     );
+};
+
+let getSnarkWorkCreatedByUser = blocks => {
+  blocks |> calculateProperty(calculateSnarkWorkCount);
+};
+
+let mapMetricsToUser = (metricsMap, metrics) => {
+  let (metricName, metricData) = metrics;
+
+  metricsMap
+  |> StringMap.fold(
+       (publicKey, metricCount, map) => {
+         map
+         |> StringMap.update(publicKey, value =>
+              switch (value) {
+              | Some(currentMetrics) =>
+                Some(
+                  Array.append(
+                    [|(metricName, metricCount)|],
+                    currentMetrics,
+                  ),
+                )
+              | None => Some([|(metricName, metricCount)|])
+              }
+            )
+       },
+       metricData,
+     );
+};
+
+/*
+     Returns a map that contains a publicKey and an array of tuples representing it's data.
+     The tuple is a pair of metric name and the metric value.
+
+     The final output should be of the form:
+       [
+         publicKey1: [(blocks_created, 2), (transactions_sent, 5)]
+         publicKey2: [(blocks_created, 1))]
+         ...
+       ]
+
+ */
+let mergeMetrics = metricsList => {
+  metricsList
+  |> Array.fold_left(
+       (metricsMap, metrics) => {mapMetricsToUser(metricsMap, metrics)},
+       StringMap.empty,
+     );
+};
+
+let delegateMetricsByType = (metric, blocks) => {
   Types.Metrics.(
-    Array.map(
-      metric => {
-        switch (metric) {
-        | BlocksCreated => blocks |> calculateBlocksCreated
-        | _ => StringMap.empty
-        }
-      },
-      metrics,
-    )
+    switch (metric) {
+    | BlocksCreated =>
+      blocks |> getBlocksCreatedByUser |> encodeMetric(Some(BlocksCreated))
+    | TransactionsSent =>
```

Comment:
I have a metric that's transactions received by each user. maybe we can rename this to TransactionsSentBy and then mine can be TransactionsSentTo ?

Reply:
@michellewong793 It seems like if you were writing a metric that did that it would make sense to call it TransactionsReceived?

Reply:
oh yeah, received works too. i was referring to the one about number of transactions sent to address